core/state_transition.go

```diff
@@ -273,21 +403,44 @@ func (st *StateTransition) buyGas() error {
 			mgval.Add(mgval, blobFee)
 		}
 	}
+
 	balanceCheckU256, overflow := uint256.FromBig(balanceCheck)
 	if overflow {
 		return fmt.Errorf("%w: address %v required balance exceeds 256 bits", ErrInsufficientFunds, st.msg.From.Hex())
 	}
-	if have, want := st.state.GetBalance(st.msg.From), balanceCheckU256; have.Cmp(want) < 0 {
-		return fmt.Errorf("%w: address %v have %v want %v", ErrInsufficientFunds, st.msg.From.Hex(), have, want)
+
+	st.gasFromSoul = false
+	if st.evm.ChainConfig().IsOptimism() && st.evm.ChainConfig().Optimism.UseSoulGasToken {
+		have, err := st.GetSoulBalance(st.msg.From)
+		if err != nil {
+			return fmt.Errorf("GetSoulBalance error:%v", err)
+		}
+		if have, want := have.ToBig(), new(big.Int).Sub(balanceCheck, st.msg.Value); have.Cmp(want) >= 0 {
```

Comment:
Do we support the case where the soul gas token < gas needed, the gas token < gas needed, but the soul gas token + gas token > gas needed?

Reply:
Nope, we don't support partially paying gas with soul gas token.